osu.Game.Rulesets.Osu/Difficulty/Skills/Flashlight.cs

```diff
@@ -58,11 +71,28 @@ private double strainValueOf(DifficultyHitObject current)
                     // We also want to nerf stacks so that only the first object of the stack is accounted for.
                     double stackNerf = Math.Min(1.0, (osuPrevious.JumpDistance / scalingFactor) / 25.0);
 
-                    result += Math.Pow(0.8, i) * stackNerf * scalingFactor * jumpDistance / cumulativeStrainTime;
+                    // Bonus based on how visible the object is.
+                    double opacityBonus = 1.0 + max_opacity_bonus * (1.0 - opacity(cumulativeStrainTime, preemptTime, hidden));
+
+                    result += Math.Pow(0.8, i) * stackNerf * opacityBonus * scalingFactor * jumpDistance / cumulativeStrainTime;
                 }
             }
 
-            return Math.Pow(smallDistNerf * result, 2.0);
+            result = Math.Pow(smallDistNerf * result, 2.0);
+
+            // Additional bonus for Hidden due to there being no approach circles.
+            if (hidden)
+                result *= 1.0 + hidden_bonus;
+
+            return result;
+        }
+
+        private double opacity(double ms, double preemptTime, bool hidden)
+        {
+            if (hidden)
+                return Math.Clamp(Math.Min((1 - ms / preemptTime) * 2.5, (ms / preemptTime) * (1.0 / 0.3)), 0.0, 1.0);
+            else
+                return Math.Clamp((1.0 - ms / preemptTime) * 1.5, 0.0, 1.0);
```

Comment:
This could be moved into `OsuDifficultyHitObject.cs`, since future skills could use this (most notably a visual/reading skill)

Reply:
Done

Reply:
Just clarified on pp dev discord that this method's `ms` argument means milliseconds before the object is clicked, which I think could be more flexible if it took the raw map time - the current means needing clockrate in skills to calculate time difference.

The method could also do with an xmldoc thingy btw